### PR TITLE
Determine binary dir independently of data dir (fixes #2891)

### LINKED ISSUE #2891: "Warning: Server executable not found, can't run server"
In the main menu, "Run server" button won't start server instance, even though it is installed in `/usr/bin/DDNet-Server`. Environment: Arch Linux 64, [ddnet] 15.0.4 (AUR)

[ddnet]: https://aur.archlinux.org/packages/ddnet/

### PR DESCRIPTION
**Pull request by def-**
@Learath2 Could you check if this can be extended to also find the DDNet-Server executable on Mac?

### CODE REVIEW

File: src/engine/shared/storage.cpp [hunk 1]
```diff
@@ -239,6 +235,65 @@ class CStorage : public IStorage
...
+					io_close(File);
+					str_copy(m_aBinarydir, aDirs[i], sizeof(m_aDatadir));
+					return;
+				}
+			}
+		}
+#endif
+
+		// no data-dir found
+		dbg_msg("storage", "warning no data directory found");
```
Comment (heinrich5991) [inline src/engine/shared/storage.cpp:295]:
```suggestion
		dbg_msg("storage", "warning no binary directory found");
```

Comment (def-) [reply src/engine/shared/storage.cpp:295]: Done.

**Comment by def-** [comment]
> @Learath2 Could you check if this can be extended to also find the DDNet-Server executable on Mac?

Done. Please test if it works.

**Comment by Learath2** [comment]
I didn't test this yet but `/../../../DDNet-Server.app/Contents/MacOS` this sounds like a huge assumption and in general Storage being a shared engine module, I don't think it's great that it's even looking for `Server`.

Is this supposed to find the directory where a server executable is? In that case I think `BinaryDir` is a misnomer. It should probably be `ServerDir`.

Now that I think about it, couldn't this accidentally find a server/client executable that's different from the one we are currently running?

**Comment by def-** [comment]
> Now that I think about it, couldn't this accidentally find a server/client executable that's different from the one we are currently running?

Yes, and it's been that way since forever. Take a random DDNet executable and run it while having DDNet installed in your package manager.

> Is this supposed to find the directory where a server executable is? In that case I think BinaryDir is a misnomer. It should probably be ServerDir.

We use BinaryDir for two things at the moment, autoupdater and running DDNet-Server. The autoupdater only works when all are in the same directory, so the Mac OSX case as well as the Linux package manager case would break it but they both don't have autoupdater.

> I didn't test this yet but /../../../DDNet-Server.app/Contents/MacOS this sounds like a huge assumption and in general Storage being a shared engine module, I don't think it's great that it's even looking for Server.

What's a better way to find it?

File: src/engine/shared/storage.cpp [hunk 2]
```diff
@@ -226,16 +223,88 @@ class CStorage : public IStorage
...
+				str_format(aBuf, sizeof(aBuf), "%s/" PLAT_SERVER_EXEC, aDirs[i]);
+				IOHANDLE File = io_open(aBuf, IOFLAG_READ);
+				if(File)
+				{
+					io_close(File);
+					str_copy(m_aBinarydir, aDirs[i], sizeof(m_aDatadir));
+					return;
+				}
+			}
+		}
```
Comment (heinrich5991) [inline src/engine/shared/storage.cpp:304]: One more thing that was also done in the other PR: If there's no path in the `argv[0]`, the server is probably also somewhere in $PATH, so we can also call it without any path.

Comment (def-) [reply src/engine/shared/storage.cpp:304]: Done.

**Comment by heinrich5991** [comment]
Looks good to me then. Thanks.

bors r+

**Pull request merged**